lndsigner.go

```diff
@@ -0,0 +1,340 @@
+	// If empty a regular network listener will be created.
+	RPCListeners []*ListenerWithSignal
+}
+
+// Main is the true entry point for lnd. It accepts a fully populated and
+// validated main configuration struct and an optional listener config struct.
+// This function starts all main system components then blocks until a signal
+// is received on the shutdownChan at which point everything is shut down again.
+func Main(cfg *Config, lisCfg ListenerCfg) error {
+	// mkErr makes it easy to return logged errors.
+	mkErr := func(format string, args ...interface{}) error {
+		signerLog.Errorf("Shutting down because error in main "+
+			"method: "+format, args...)
+		return fmt.Errorf(format, args...)
+	}
+
+	var network string
+	switch {
+	/*case cfg.MainNet:
+	network = "mainnet"
+	*/
+	case cfg.TestNet3:
+		network = "testnet"
+
+	case cfg.SimNet:
+		network = "simnet"
+
+	case cfg.RegTest:
+		network = "regtest"
+
+	case cfg.SigNet:
+		network = "signet"
+	}
+
+	signerLog.Infof("Active chain: %v (network=%v)", "bitcoin", network)
+
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Use defaults for vault client, including getting config from env.
+	vaultClient, err := api.NewClient(nil)
+	if err != nil {
+		return mkErr("error creating vault client: %v", err)
+	}
+
+	signerClient := vaultClient.Logical()
+
+	listAcctsResp, err := signerClient.ReadWithData(
+		"lndsigner/lnd-nodes/accounts",
+		map[string][]string{
+			"node": []string{cfg.NodePubKey},
+		},
+	)
+	if err != nil {
+		return mkErr("error listing accounts: %v", err)
+	}
+
+	acctList, ok := listAcctsResp.Data["acctList"]
+	if !ok {
+		return mkErr("accounts not returned")
+	}
+
+	err = os.WriteFile(
+		cfg.OutputAccounts,
+		[]byte(acctList.(string)),
+		outputFilePermissions,
+	)
+	if err != nil {
+		return mkErr("error writing account list: %v",
+			err)
+	}
+
+	serverOpts, err := getTLSConfig(cfg)
+	if err != nil {
+		return mkErr("unable to load TLS credentials: %v", err)
+	}
+
+	// If we have chosen to start with a dedicated listener for the
+	// rpc server, we set it directly.
+	grpcListeners := append([]*ListenerWithSignal{}, lisCfg.RPCListeners...)
+	if len(grpcListeners) == 0 {
+		// Otherwise we create listeners from the RPCListeners defined
+		// in the config.
+		for _, grpcEndpoint := range cfg.RPCListeners {
+			// Start a gRPC server listening for HTTP/2
+			// connections.
+			lis, err := ListenOnAddress(grpcEndpoint)
+			if err != nil {
+				return mkErr("unable to listen on %s: %v",
+					grpcEndpoint, err)
+			}
+			defer lis.Close()
+
+			grpcListeners = append(
+				grpcListeners, &ListenerWithSignal{
+					Listener: lis,
+					Ready:    make(chan struct{}),
+				},
+			)
+		}
+	}
+
+	// Initialize the rpcServer and add its interceptor to the server
+	// options.
+	rpcServer := newRPCServer(cfg, signerClient)
+	serverOpts = append(
+		serverOpts,
+		grpc.ChainUnaryInterceptor(rpcServer.intercept),
```

Comment:
Macaroon interceptor redundant now?

Reply:
Interceptor is still used for policy enforcement but I've removed all the macaroon stuff.

Reply:
How is it used for policy enforcement? In the current revision I see that it just attached the keyring to the context?

Reply:
I've just rebased the branch I'm using to develop policy enforcement on this code here:

https://github.com/bottlepay/lndsigner/blob/offchain-ratelimiting/rpcserver.go#L80

Just finishing up the actual rate-limiting for tomorrow's demo now.